drivers/ws281x/systick_gpio_ll.c

```diff
@@ -0,0 +1,134 @@
+#include "debug.h"
+
+/* (+ NS_PER_SEC - 1): Rounding up, as T1H is the time that needs to distinctly
+ * longer than T0H.
+ *
+ * Then adding +1 extra, because the spin loop adds another layer of jitter. A
+ * more correct version would be to add the spin loop time before rounding (and
+ * then rounding up), but as that time is not available, spending one more
+ * cycle is the next best thing to do. */
+const int ticks_one = ((uint64_t)WS281X_T_DATA_ONE_NS * WS281X_TIMER_FREQ + NS_PER_SEC - 1)
+                    / NS_PER_SEC + 1;
+/* Rounding down, zeros are better shorter */
+const int ticks_zero = (uint64_t)WS281X_T_DATA_ZERO_NS * (uint64_t)WS281X_TIMER_FREQ / NS_PER_SEC;
+/* No particular known requirements, but we're taking longer than that anyway
+ * because we don't clock the times between bits. */
+const int ticks_data = (uint64_t)WS281X_T_DATA_NS * (uint64_t)WS281X_TIMER_FREQ / NS_PER_SEC;
+
+static void _systick_start(uint32_t ticks)
+{
+    /* disable SysTick, clear value */
+    SysTick->CTRL = 0;
+    SysTick->VAL = 0;
+    /* prepare value in re-load register */
+    SysTick->LOAD = ticks;
+    /* start and wait for the load value to be applied */
+    SysTick->CTRL = SysTick_CTRL_CLKSOURCE_Msk | SysTick_CTRL_ENABLE_Msk;
+    while (SysTick->VAL == 0) { /* Wait for SysTick to start and spin */ }
+}
+
+static void _systick_wait(void)
+{
+    while (!(SysTick->CTRL & SysTick_CTRL_COUNTFLAG_Msk)) { /* busy wait */ }
+}
+
+void ws281x_write_buffer(ws281x_t *dev, const void *buf, size_t size)
+{
+    assert(dev);
+
+    /* the high time for one can be as high as 5 seconds in practise, so
+     * rather be on the high side by adding a few CPU cycles. */
+    const uint32_t ticks_one = DIV_ROUND_UP((uint64_t)WS281X_T_DATA_ONE_NS * (uint64_t)coreclk(), NS_PER_SEC) + 16;
+    /* the low time should rather be on the short side, so rounding down */
+    const uint32_t ticks_zero = (uint64_t)(WS281X_T_DATA_ZERO_NS - 50) * (uint64_t)coreclk() / NS_PER_SEC;
+    /* the remaining time doesn't matter to much, should only be enough for the
+     * LEDs to detect the low phase. And not way to much to be detected as
+     * reset */
+    const uint32_t ticks_bit = DIV_ROUND((uint64_t)WS281X_T_DATA_NS * (uint64_t)coreclk(), NS_PER_SEC);
+
+    const uint8_t *pos = buf;
+    const uint8_t *end = pos + size;
+
+    gpio_port_t port = gpio_get_port(dev->params.pin);
+    uword_t mask = 1U << gpio_get_pin_num(dev->params.pin);
+
+    unsigned irq_state = irq_disable();
+    while (pos < end) {
+        uint8_t data = *pos;
+        for (uint8_t cnt = 8; cnt > 0; cnt--) {
+            uint32_t ticks_high = (data & 0x80) ? ticks_one : ticks_zero;
+            uint32_t ticks_low = ticks_bit - ticks_high;
+            _systick_start(ticks_high);
+            gpio_ll_set(port, mask);
+            _systick_wait();
+            gpio_ll_clear(port, mask);
+            _systick_start(ticks_low);
+            _systick_wait();
+            data <<= 1;
+        }
```

Comment:
lets try to rewrite my pseudo code in pseudo c

```suggestion
    _systick_start( short );
        for (uint8_t cnt = 8; cnt > 0; cnt--) {
            int bit = !!(data & 0x80);
            gpio_ll_set(port, mask);
            if(!bit){
                _systick_wait();
            }else{
                _systick_wait();_systick_wait();
            }
            gpio_ll_clear(port, mask);
            if(!bit){
                _systick_wait();_systick_wait();
            }else{
                _systick_wait();
            }
            data <<= 1;
        }
```

Reply:
and then there is the potential use of the systick irq
```
global byte
global count

systick_irq(){
if(count % 4 = 0)  gpio_set();
else if (count % 4 =1 && !(byte & 0x80 >> (count / 4))) gpio_clr;
else if (count % 4 = 2 ) gpio_clr;
count++;
)
```

Reply:
The overhead of entering the IRQ, backing up the state etc. takes way longer than the total time budget

Reply:
ok the irq thing might not be the right choice if you got 4 cycles